src/cruise_control/scripts/pid_controller.py

```diff
@@ -0,0 +1,93 @@
+import math
+import time
+import os
+from enum import Enum
+
+# The output of the PID controller is always a positive number after being
+# translated to a value that AirSim can use for acceleration or braking.
+# Thus a distinction needs to be made if the returned value is for accelerating or braking
+class ThrottleAction(Enum):
+    Accelerate = 0
+    Brake = 1
+
+class PIDController:
+
+    def __init__(self):
+        # This depends on exp_factor in pidToCarValues()
+        self.speedAdjustmentFactor = 1.15
+        self.currentSpeed = 0.0
+        self.targetSpeed = 5.0 * self.speedAdjustmentFactor
+
+        self.speed_difference_sum = 0
+        self.lastError = 0
+        self.speedDifference = 0
+
+        pidCoefficientsFile = open(os.path.abspath(os.path.dirname(__file__)) + "/pid_values.txt")
+
+        # Used in integral windup mitigation. See publish_results()
+        self.potentialKiValue = float(pidCoefficientsFile.readline())
+        self.clampValue = 0.995  # Must be in range of [0, 1)
+
+        self.kP = float(pidCoefficientsFile.readline())
+        self.kI = self.potentialKiValue
+        self.kD = float(pidCoefficientsFile.readline())
+
+    # Converts the PID controller output value in the range of [0, 1) that the throttle and brake can have
+    def pidToCarValues(self, inp):
+
+        # A lower value means a more gradual throttling and acceleration. However a higher input is needed to get a
+        # output value that is needed for a car to achieve its target speed than the PID controller will output. To deal
+        # with this, make the target speed higher than what is desired internally using self.speedAdjustmentFactor
+        exp_factor = -1
```

Comment:
exp_factor can have its value changed based off of what is observed to be better (have to change self.SpeedAdjustmentFactor as well in that case).

With factor = -0.5 I found that at higher speeds, say 7 m/s, there was less of a harsh brake of the speed during the initial acceleration when the speed went above desired, but in the last turn our default test path it seemed to slow down rather heavily at times (to around 6.6 m/s). With factor = -2 the opposite was true. I'd say leave it for now and once we have more test paths maybe reconsider changing this value if we want.

Theoretically only the kP, kI and kD coefficients should be changed so that should be noted, though that also depends on having a good translation function between PID output to throttle value.